### PR TITLE
Bpe vocab

### PR DESCRIPTION
**Pull request by monsieurzhang**
support vocab

### CODE REVIEW

File: cli/tokenize.cc [hunk 1]
```diff
@@ -77,10 +79,19 @@ int main(int argc, char* argv[])
   }
 #endif
 
+  std::string bpe_vocab_path;
```
Comment (guillaumekln) [inline cli/tokenize.cc]:
```cpp
std::string bpe_vocab_path = vm["bpe_vocab"].as<std::string>();
```

File: cli/tokenize.cc [hunk 2]
```diff
@@ -77,10 +79,19 @@ int main(int argc, char* argv[])
...
+  if (!vm["bpe_vocab"].as<std::string>().empty())
+    bpe_vocab_path = vm["bpe_vocab"].as<std::string>();
+
+  int bpe_vocab_threshold;
```
Comment (guillaumekln) [inline cli/tokenize.cc]:
```cpp
int bpe_vocab_threshold = vm["bpe_vocab_threshold"].as<int>();
```

File: include/onmt/BPE.h [hunk 3]
```diff
@@ -16,17 +17,33 @@ namespace onmt
 
     std::vector<std::string> encode(const std::string& str) const override;
 
+    void init_bpe_vocab(std::string path, int bpe_vocab_threshold);
```
Comment (guillaumekln) [inline include/onmt/BPE.h]:
```cpp
const std::string& path
```

File: include/onmt/BPE.h [hunk 4]
```diff
@@ -16,17 +17,33 @@ namespace onmt
...
+    void set_joiner(std::string joiner)
```
Comment (guillaumekln) [inline include/onmt/BPE.h]: What about spacers?

Would it make sense to pass the joiner and spacer as a list to `init_bpe_vocab` which will remove such substrings in each word when creating `_bpe_vocab`?

File: src/BPE.cc [hunk 5]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+      out.push_back(segment);
+    }
+    else
+    {
+      recursive_split(segment, out, false);
+    }
+
+  }
+
+  void BPE::recursive_split_right(std::string & segment, std::vector<std::string> & out, bool finalflag) const
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
const std::string& segment
```

File: src/Tokenizer.cc [hunk 6]
```diff
@@ -81,6 +83,11 @@ namespace onmt
 #endif
     {
       set_bpe_model(model_path, _cache_model);
+      if (_subword_encoder != nullptr && !bpe_vocab_path.empty())
+      {
+        ((BPE *)_subword_encoder)->init_bpe_vocab(bpe_vocab_path, bpe_vocab_threshold);
```
Comment (guillaumekln) [inline src/Tokenizer.cc]:
```cpp
dynamic_cast<BPE*>(_subword_encoder)
```

File: src/BPE.cc [hunk 7]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+      recursive_split_right(right, out, finalflag);
+    }
+    else
+    {
+      out.push_back(segment);
+    }
+
+  }
+
+  void BPE::recursive_split_left(std::string & segment, std::vector<std::string> & out) const
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
const std::string& segment
```

File: src/BPE.cc [hunk 8]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+    else
+    {
+      recursive_split(segment, out, false);
+    }
+
+  }
+
+  void BPE::recursive_split_right(std::string & segment, std::vector<std::string> & out, bool finalflag) const
+  {
+    if((finalflag && _bpe_vocab.find(segment) != _bpe_vocab.end()) || (!finalflag && _bpe_vocab.find(segment + _joiner) != _bpe_vocab.end()))
```
Comment (guillaumekln) [inline src/BPE.cc]: Use `_bpe_vocab.count(...) > 0` instead.

File: src/BPE.cc [hunk 9]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+    else
+    {
+      out.push_back(segment);
+    }
+
+  }
+
+  void BPE::recursive_split_left(std::string & segment, std::vector<std::string> & out) const
+  {
+    if (_bpe_vocab.find(segment + _joiner) != _bpe_vocab.end())
```
Comment (guillaumekln) [inline src/BPE.cc]: Use `_bpe_vocab.count(...) > 0` instead.

File: src/BPE.cc [hunk 10]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+    }
+
+  }
+
+  void BPE::recursive_split(std::string & segment, std::vector<std::string> & out, bool finalflag) const
+  {
+    // Recursively split segment into smaller units (by reversing BPE merges)
+    // until all units are either in - vocabulary, or cannot be split futher.
+
+    std::unordered_map<std::string, std::pair<std::string, std::string> >::const_iterator got = _codes_reverse.find(finalflag == true ? segment+_end_of_word: segment);
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
auto got = ...
```

File: src/BPE.cc [hunk 11]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+      }
+      else
+      {
+        recursive_split(segment, out, std::next(it) == orig.end());
+      }
+    }
+
+  }
+
+  void BPE::recursive_split(std::string & segment, std::vector<std::string> & out, bool finalflag) const
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
const std::string& segment
```

File: src/BPE.cc [hunk 12]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+      size_t sep = line.find(' ');
+      if (sep != std::string::npos)
+      {
+        if(std::stoi(line.substr(sep + 1)) >= bpe_vocab_threshold)
+          _bpe_vocab.emplace(line.substr(0, sep));
+      }
+    }
+  }
+
+  void BPE::check_vocab_and_split(std::vector<std::string> & orig, std::vector<std::string> & out) const
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
const std::vector<std::string>& orig
```

File: src/BPE.cc [hunk 13]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+  {
+    // Check for each segment in word if it is in-vocabulary,
+    // and segment OOV segments into smaller units by reversing the BPE merge operations
+
+    for (std::vector<std::string>::iterator it = orig.begin(); it != orig.end(); ++it)
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
for (auto it = ...)
```

File: src/BPE.cc [hunk 14]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+    {
+      std::string segment = *it;
```
Comment (guillaumekln) [inline src/BPE.cc]:
```cpp
const std::string& segment = ...
```

File: src/BPE.cc [hunk 15]
```diff
@@ -198,4 +222,94 @@ namespace onmt
...
+
+      // if it is not the last, add joiner
+      if (_bpe_vocab.find(std::next(it) == orig.end() ? segment : segment+_joiner) != _bpe_vocab.end())
```
Comment (guillaumekln) [inline src/BPE.cc]: Use `_bpe_vocab.count(...) > 0` instead.

File: src/BPE.cc [hunk 16]
```diff
@@ -1,5 +1,5 @@
 #include "onmt/BPE.h"
-
+#include <iostream>
```
Comment (guillaumekln) [inline src/BPE.cc]: This is not used.

File: cli/tokenize.cc (hunk 1 above)
Comment (monsieurzhang) [reply cli/tokenize.cc]: fix

File: cli/tokenize.cc (hunk 2 above)
Comment (monsieurzhang) [reply cli/tokenize.cc]: fix

File: include/onmt/BPE.h (hunk 3 above)
Comment (monsieurzhang) [reply include/onmt/BPE.h]: fix

File: src/BPE.cc (hunk 16 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 12 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 13 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 14 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 15 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 11 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 10 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 7 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 9 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 5 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/BPE.cc (hunk 8 above)
Comment (monsieurzhang) [reply src/BPE.cc]: fix

File: src/Tokenizer.cc (hunk 6 above)
Comment (monsieurzhang) [reply src/Tokenizer.cc]: fix

**Pull request merged**